src/chainstate/burn/db/burndb.rs

```diff
@@ -1001,61 +1055,70 @@ where
 
     /// Find out whether or not a given consensus hash is "recent" enough to be used in the
     /// canonical burnchain history
-    pub fn is_fresh_consensus_hash(conn: &Connection, current_block_height: u64, consensus_hash: &ConsensusHash) -> Result<bool, db_error> {
+    pub fn is_fresh_consensus_hash(conn: &Connection, current_block_height: u64, consensus_hash_lifetime: u32, consensus_hash: &ConsensusHash) -> Result<bool, db_error> {
         if current_block_height > ((1 as u64) << 63) - 1 {
             return Err(db_error::TypeError);
         }
 
-        if current_block_height < (CONSENSUS_HASH_LIFETIME as u64) {
+        if current_block_height < (consensus_hash_lifetime as u64) {
             return Err(db_error::TypeError);
         }
 
         let qry = "SELECT COUNT(consensus_hash) FROM snapshots WHERE canonical = 1 AND consensus_hash = ?1 AND block_height >= ?2 AND block_height <= ?3";
-        let args = [&consensus_hash.to_hex(), &((current_block_height as u32) - CONSENSUS_HASH_LIFETIME) as &ToSql, &(current_block_height as u32) as &ToSql];
+        let args = [&consensus_hash.to_hex(), &((current_block_height as u32) - consensus_hash_lifetime) as &ToSql, &(current_block_height as u32) as &ToSql];
         let count = BurnDB::<A, K>::query_count(conn, &qry.to_string(), &args)?;
         Ok(count != 0)
     }
 
-    /// Determine whether or not a leader key is available to be consumed by a block commitment at the given block height.
-    /// That is, there must _not_ be a block commit paired with this leader key already.
-    pub fn is_leader_key_available(conn: &Connection, leader_key: &LeaderKeyRegisterOp<A, K>) -> Result<bool, db_error> 
+    /// Determine whether or not a leader key has been consumed by a subsequent block commitment.
+    /// Will return false if the leader key does not exist.
+    pub fn is_leader_key_consumed(conn: &Connection, leader_key: &LeaderKeyRegisterOp<A, K>) -> Result<bool, db_error> 
     where
         LeaderBlockCommitOp<A, K>: FromRow<LeaderBlockCommitOp<A, K>> + RowOrder
     {
         if leader_key.block_number > ((1 as u64) << 63) - 1 {
             return Err(db_error::TypeError);
         }
 
-        let row_order = "block_commits.txid,block_commits.burn_header_hash";
-        let qry = format!("SELECT COUNT({}) FROM block_commits JOIN history ON block_commits.txid = history.txid AND block_commits.burn_header_hash = history.burn_header_hash \
-                          WHERE history.canonical = 1 AND block_commits.block_height - block_commits.key_block_backptr = ?1 AND block_commits.key_vtxindex = ?2", row_order);
+        let qry = "SELECT COUNT(*) FROM block_commits JOIN history ON block_commits.txid = history.txid AND block_commits.burn_header_hash = history.burn_header_hash \
+                   WHERE history.canonical = 1 AND block_commits.block_height - block_commits.key_block_backptr = ?1 AND block_commits.key_vtxindex = ?2".to_string();
```

Comment:
the reason for representing the key_block_backptr this way is to reduce the bits required to identify the key_block ?

and this _isn't_ a problem, because block commits _already_ must be broadcast within a single block to actually be candidates for an epoch?

Reply:
Yes -- we don't have enough space for a 4-byte value.  Also, a block commit must be sent _strictly after_ its leader's public key commitment, and all block commits in the same sortition must be sent in the same block.